### PR TITLE
Feature/set weights and bias

### PR DESCRIPTION
**Pull request by Wissiak**
@PetrKubes97 Thanks for your review! I'm very happy that you like my proposal and I've implemented the changes you've asked for. 

The diff for index.html still is a mess but somehow if I reset the file and add the changes manually, WebStorm doesn't recognize the space / line break changes 😕

### CODE REVIEW

File: src/neuralNetwork/Connection.ts [hunk 1]
```diff
@@ -1,14 +1,15 @@
...
+  private weight: number;
   private inputNeuron: Neuron;
   private outputNeuron: Neuron;
   private sampleWeightChanges: number[] = [];
 
-  constructor(input: Neuron, output: Neuron) {
+  constructor(input: Neuron, output: Neuron, weight?: number) {
     this.inputNeuron = input;
     this.outputNeuron = output;
+    this.weight = weight || Math.random();
```
Comment (PetrKubes97) [inline src/neuralNetwork/Connection.ts]: I think the weights should still be between centered around zero. Not sure why I had the * 10 - 5 before, but at least -0.5 would make sense.

File: src/neuralNetwork/NeuralCore.ts [hunk 2]
```diff
@@ -8,6 +8,7 @@ export class NeuralCore {
   private outputSize: number;
 
   private layerCnt: number;
+  private weightList: number[][][];
```
Comment (PetrKubes97) [inline src/neuralNetwork/NeuralCore.ts]: Weight is saved in the neuron's connection, so this should not be necessary.

File: src/neuralNetwork/NeuralCore.ts [hunk 3]
```diff
@@ -66,8 +67,42 @@ export class NeuralCore {
...
+      }
+    }
+    for (let i = 0; i < weights[weights.length - 1].length; i++) {
+      if (weights[weights.length - 1][i].length !== this.hiddenLayerSizes[this.hiddenLayerSizes.length - 1] + 1) { //+1 for bias
+        throw `Weights at output layer of neuron ${i+1} do not match the input count`;
+      }
+    }
 
+    this.weightList = weights;
+  }
```
Comment (PetrKubes97) [inline src/neuralNetwork/NeuralCore.ts:123]: This functions should loop through all the neurons and their connections and set the weights accordingly.

File: src/neuralNetwork/NeuralCore.ts [hunk 4]
```diff
@@ -331,17 +366,32 @@ export class NeuralCore {
...
+        }
+        this.neurons[l].forEach((currNeuron, fromIdx) => {
+          let weight;
+          if (weights.length > fromIdx + 1) {
+            // Do not use last weight since it corresponds to the bias
+            weight = weights[fromIdx]
+          } else {
+            // Use default value from Connection
+          }
+          const connection = new Connection(currNeuron, nextNeuron, weight)
```
Comment (PetrKubes97) [inline src/neuralNetwork/NeuralCore.ts]: This should be removed after the previous changes.

**Review by PetrKubes97 — CHANGES_REQUESTED** [review]

**Comment by PetrKubes97** [comment]
Do you think it is necessary to have the set weights in a separate variable from the actual weights (Neuron->connections)? I might be missing something, but seems more intuitive to me to just set actual weights. 

For the formatting, I'll do a reformat and add .editorconfig later today.

**Comment by Wissiak** [comment]
Your observation is correct and I've adapted the code to set the weights directly on the connection.

**Comment by PetrKubes97** [comment]
I've pushed the .editorconfig file and reformatted the whole project. I'm using intellij, so the formatting should match.

**Comment by Wissiak** [comment]
I have updated the pull-request now with the .editorconfig and the diff looks better now 👍

**Comment by PetrKubes97** [comment]
Ok, looks good, however I've still found a few problems.

1. Setting random weights clears the weights field to '[]' which is not consistent with the behaviour for inputs. If input changes (using the sliders), the set input field does not reset.
2. Clicking on row from training data to activate inputs stopped working.

**Comment by Wissiak** [comment]
1. I've adapted the behavior to not reset the weights
2. The problem was that I've put the `draw()` call at the end of the `updateUI()` function. Since upon updating the UI, the cost is being calculated, `NeuralCore.evaluate(inputSample)` is being called for each training data input. The `NeuralCore.evaluate(inputSample)` function updates the input neurons to have the values of the given inputSample which then ended up showing only the last training data input. Upon reverting my `draw()` call reorder change, the UI is being updated to the right time again. 

In addition:
Since I din't know that I could change the input sample by clicking on a row in the table, I've added a hint to the UI.

**Comment by PetrKubes97** [comment]
Nice work! Merging now.

**Pull request merged**